power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -12,38 +12,23 @@
 namespace power_grid_model::math_solver {
 
 namespace detail {
-
-template <symmetry_tag sym>
-std::tuple<Idx, Idx> count_voltage_sensors(Idx const n_bus, MeasuredValues<sym> const& measured_values) {
-    Idx n_voltage_sensor{};
-    Idx n_voltage_phasor_sensor{};
-    for (Idx bus = 0; bus != n_bus; ++bus) {
-        if (measured_values.has_voltage(bus)) {
-            n_voltage_sensor++;
-            if (measured_values.has_angle_measurement(bus)) {
-                n_voltage_phasor_sensor++;
-            }
-        }
-    }
-    return std::make_tuple(n_voltage_sensor, n_voltage_phasor_sensor);
-}
-
-// count flow sensors into integer matrix with ybus structure
-// lower triangle part is always zero
-// for diagonal part, it will be one if there is bus injection
-// for upper triangle part, it will be one if there is branch flow sensor and the branch is fully connected
-// return a pair of
+// count flow and voltage phasor sensors for the observability check
+// use the ybus structure
+// the lower triangular part is always zero
+// the diagonal part will be one if there is bus injection or a voltage phasor sensor
+// the upper triangle part will be one if there is branch flow sensor and the branch is fully connected
+// return a tuple of
 //      a vector of flow sensor count
+//      a vector of voltage phasor sensor count
 //      a boolean indicating if the system is possibly ill-conditioned
 template <symmetry_tag sym>
-std::pair<std::vector<int8_t>, bool> count_flow_sensors(MeasuredValues<sym> const& measured_values,
-                                                        MathModelTopology const& topo,
-                                                        YBusStructure const& y_bus_structure) {
+std::tuple<std::vector<int8_t>, std::vector<int8_t>, bool>
```

Comment:
can you make this return type a named object? e.g.
```cpp
struct ObservingSensorCount {
    std::vector<int8_t> flow_sensor_count;
    std::vector<int8_t> voltage_phasor_count;
    bool maybe_ill_conditioned{false};
};
```

or alternatively (but maybe overloads or static accessors are needed for `const` access)

```cpp
struct ObservingSensorCount {
    std::tuple<std::vector<int8_t>, std::vector<int8_t>, bool> data{};
    constexpr decltype(auto) flow_sensor_count() { return std::get<0>(data); }
    constexpr decltype(auto) voltage_phasor_sensor_count() { return std::get<1>(data); }
    constexpr bool maybe_ill_conditioned() { return std::get<2>(data); }
};
```

I'm personally not a fan of unnamed tuples